src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs

```diff
@@ -2532,160 +2653,246 @@ var projectUniqueNamesForBuildIntToUpdate
             var now = DateTimeOffset.UtcNow;
             void cacheModifier(SourceCacheContext cache) => cache.MaxAge = now;
 
-            // If the lock file does not exist, restore before starting the operations
-            if (originalLockFile == null)
+            // Add all enabled sources for the existing projects
+            var enabledSources = SourceRepositoryProvider.GetRepositories();
+            var allSources = new HashSet<SourceRepository>(enabledSources, new SourceRepositoryComparer());
+
+            foreach (var buildIntegratedProject in buildIntegratedProjects)
             {
-                var originalRestoreResult = await DependencyGraphRestoreUtility.PreviewRestoreAsync(
-                    SolutionManager,
-                    buildIntegratedProject,
-                    originalPackageSpec,
-                    dependencyGraphContext,
-                    providerCache,
-                    cacheModifier,
-                    sources,
-                    nuGetProjectContext.OperationId,
-                    logger,
-                    token);
+                NuGetProjectAction[] nuGetProjectActions;
```

Comment:
Why is this variable needed? 
Can the line https://github.com/NuGet/NuGet.Client/blob/24c713cd927fe4920ec5ecc6fe98033f84eab727/src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs#L2673

be 
nugetProjectActionsLookup[buildIntegratedProject.MSBuildProjectPath] = new[] { nugetAction };